packages/ember-metal/lib/meta.js

```diff
@@ -0,0 +1,287 @@
+ peekValues, clearValues, writableDeps, readableDeps, getAllDeps
+ writableChainWatchers, readableChainWatchers, writableChains,
+ readableChains
+
+*/
+let members = {
+  cache: ownMap,
+  watching: inheritedMap,
+  mixins: inheritedMap,
+  bindings: inheritedMap,
+  values: inheritedMap,
+  deps: inheritedMapOfMaps,
+  chainWatchers: ownCustomObject,
+  chains: inheritedCustomObject
+};
+
+let memberNames = Object.keys(members);
+
+function Meta(obj, parentMeta) {
+  // preallocate a slot for each member
+  for (let i = 0; i < memberNames.length; i++) {
+    this[memberProperty(memberNames[i])] = undefined;
+  }
+
+  // used only internally
+  this.source = obj;
+
+  // when meta(obj).proto === obj, the object is intended to be only a
+  // prototype and doesn't need to actually be observable itself
+  this.proto = undefined;
+
+  // The next meta in our inheritance chain. We (will) track this
+  // explicitly instead of using prototypical inheritance because we
+  // have detailed knowledge of how each property should really be
+  // inherited, and we can optimize it much better than JS runtimes.
+  this.parent = parentMeta;
+
+  this._initializeListeners();
+}
+
+for (let name in listenerMethods) {
+  Meta.prototype[name] = listenerMethods[name];
+}
+
+(function setupMembers() {
+  for (let i = 0; i < memberNames.length; i++) {
+    let name = memberNames[i];
+    let implementation = members[name];
+    implementation(name, Meta);
+  }
+})();
+
+
+// Implements a member that is a lazily created, non-inheritable
+// POJO.
+function ownMap(name, Meta) {
+  let key = memberProperty(name);
+  let capitalized = capitalize(name);
+  Meta.prototype['writable' + capitalized] = function() {
+    return getOrCreateOwnMap.call(this, key);
+  };
+  Meta.prototype['readable' + capitalized] = function() { return this[key]; };
+}
+
+function getOrCreateOwnMap(key) {
+  let ret = this[key];
+  if (!ret) {
+    ret = this[key] = {};
+  }
+  return ret;
+}
+
+// Implements a member that is a lazily created POJO with inheritable
+// values.
+function inheritedMap(name, Meta) {
+  let key = memberProperty(name);
+  let capitalized = capitalize(name);
+
+  Meta.prototype['writable' + capitalized] = function() {
+    return getOrCreateInheritedMap.call(this, key);
+  };
+
+  Meta.prototype['readable' + capitalized] = function() {
+    return getInherited.call(this, key);
+  };
+
+  Meta.prototype['peek' + capitalized] = function(subkey) {
+    let map = getInherited.call(this, key);
+    if (map) {
+      return map[subkey];
+    }
+  };
+
+  Meta.prototype['clear' + capitalized] = function() {
+    this[key] = {};
+  };
+}
+
+function getOrCreateInheritedMap(key) {
+  let ret = this[key];
+  if (!ret) {
+    if (this.parent) {
+      ret = this[key] = Object.create(getOrCreateInheritedMap.call(this.parent, key));
+    } else {
+      ret = this[key] = {};
+    }
+  }
+  return ret;
+}
+
+function getInherited(key) {
+  let pointer = this;
+  while (pointer) {
```

Comment:
`pointer !== undefined` skips the coercion step, which I don't believe gets optimized out yet :(